### PR TITLE
Classification loss modifier

### PR DESCRIPTION
**Pull request by Kag-Bar**
### What
Added configurable classification loss types beyond the default cross-entropy, enabling better handling of class imbalance and stronger inter-class separation. Supported losses:

- **Weighted Cross-Entropy** — user-defined per-class importance weights
- **Label Smoothing** — regularization to prevent overconfident predictions (combinable with any loss)
- **Focal Loss** — down-weights easy examples, focuses on hard ones
- **Class-Balanced Focal Loss** — auto-computes effective-number-of-samples weights from training data
- **ArcFace Loss** — additive angular margin loss for highly discriminative feature embeddings

All options are backward compatible — default behavior is unchanged (plain cross-entropy).

### How to Use

# Weighted CE — boost "Reject" class
yolo classify train model=yolo11n-cls.pt data=my_data cls_loss=ce class_weights="{Reject: 10.0}"

# Focal loss
yolo classify train model=yolo11n-cls.pt data=my_data cls_loss=focal focal_gamma=2.0

# Class-balanced focal (auto frequency weights)
yolo classify train model=yolo11n-cls.pt data=my_data cls_loss=cb_focal cb_beta=0.999

# ArcFace with angular margin
yolo classify train model=yolo11n-cls.pt data=my_data cls_loss=arcface arcface_margin=0.5 arcface_scale=30.0

# Combine: ArcFace + class weights + label smoothing
yolo classify train model=yolo11n-cls.pt data=my_data cls_loss=arcface class_weights="{Reject: 10.0}" label_smoothing=0.1

### Config Parameters (default.yaml)

Parameter | Default | Description
-- | -- | --
cls_loss | ce | Loss type: ce, focal, cb_focal, arcface
class_weights | None | Dict or list of per-class importance weights
label_smoothing | 0.0 | Label smoothing factor (works with any loss)
focal_gamma | 2.0 | Focal loss focusing parameter γ
cb_beta | 0.9999 | Class-balanced effective-number β
arcface_margin | 0.5 | ArcFace angular margin (radians)
arcface_scale | 30.0 | ArcFace logit scale factor

### File Changes

File | Change
-- | --
ultralytics/cfg/default.yaml | Added cls_loss, label_smoothing, focal_gamma, cb_beta, arcface_margin, arcface_scale parameters
ultralytics/cfg/__init__.py | Registered new params in CFG_FLOAT_KEYS, CFG_FRACTION_KEYS, and allowed_custom_keys for validation
ultralytics/utils/loss.py | Refactored v8ClassificationLoss to support all 5 loss types with _focal_loss() and _arcface_loss() methods
ultralytics/nn/modules/head.py | Added _return_features flag to Classify head — returns raw features (instead of logits) during training when ArcFace is active
ultralytics/nn/tasks.py | Updated ClassificationModel.init_criterion() to configure head and loss based on cls_loss, with safe reset of _return_features on loss-type switch
ultralytics/models/yolo/classify/train.py | Added _resolve_class_weights() and _compute_class_counts() to ClassificationTrainer for weight/count resolution from dataset
tests/test_classification_loss.py | 23 unit tests covering all loss types, backward compatibility, ArcFace eval fallback, and model reuse scenarios
scripts/test_classification_loss.py | Demo script comparing 12 loss configurations on a real dataset

## Summary by CodeRabbit

* **New Features**
  * Multiple classification losses supported (CE, Focal, Class‑Balanced Focal, ArcFace) with configurable label smoothing, focal gamma, cb_beta, ArcFace margin/scale, and per‑class weighting; ArcFace training/eval integrated.
  * Default configs updated to expose these options; optional Weights & Biases integration enabled.

* **New Tools**
  * Experimental script to run and compare classification-loss training configurations and produce comparative metrics.

* **Tests**
  * Comprehensive test suite validating loss behaviors, weighting, ArcFace modes, and edge cases.

* **Other**
  * Added /data to ignore list and minor CLI/output formatting tweaks.

### CODE REVIEW

File: scripts/find_bad_sample.py [hunk 1]
```diff
@@ -1,22 +1,23 @@
 """Script to find problematic samples in GPU 3's partition."""
 
-import sys
 import signal
 from pathlib import Path
 
+
 # Timeout handler
 class TimeoutError(Exception):
     pass
```
[the remark this answers is not in the document]
Comment (Kag-Bar) [reply scripts/find_bad_sample.py:9]: ignoring scripts/

File: scripts/test_classification_loss.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+  1. ce              – standard cross-entropy (baseline)
+  2. ce + w10        – weighted CE (Reject ×10)
+  3. ce + w20        – weighted CE (Reject ×20)
+  4. focal γ=2       – focal loss
+  5. focal γ=5       – focal loss with stronger hard-example focus
+  6. focal γ=2 + w10 – focal loss + heavy Reject weight
+  7. cb_focal γ=2    – class-balanced focal (auto frequency weights)
+  8. cb_focal + w10  – class-balanced focal + extra Reject weight
+  9. ce + smoothing  – cross-entropy with label smoothing
+
```
[the remark this answers is not in the document]
Comment (Kag-Bar) [reply scripts/test_classification_loss.py:16]: ignoring scripts/

File: scripts/test_classification_loss.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    split_dir.mkdir(parents=True, exist_ok=True)
+
+    classes = sorted(d.name for d in raw_dir.iterdir() if d.is_dir() and d.name not in ("train", "val", "test"))
+    for cls_name in classes:
+        cls_dir = raw_dir / cls_name
+        images = sorted(f for f in cls_dir.iterdir() if f.is_file())
+        random.shuffle(images)
+        n_val = max(1, int(len(images) * val_ratio))
+        val_imgs = images[:n_val]
+        train_imgs = images[n_val:]
```
[the remark this answers is not in the document]
Comment (Kag-Bar) [reply scripts/test_classification_loss.py:58]: ignoring scripts/

File: ultralytics/models/yolo/classify/train.py [hunk 4]
```diff
@@ -65,8 +65,71 @@ def __init__(self, cfg=DEFAULT_CFG, overrides: dict[str, Any] | None = None, _ca
...
+
+        # Build a temporary dataset to count samples
+        dataset = self.build_dataset(str(train_path), mode="train")
+        for sample in dataset.samples:
+            cls_idx = sample[1]
+            if cls_idx < nc:
+                counts[cls_idx] += 1
+
+        self.args.class_counts = counts
+        LOGGER.info(f"Classification class counts (train): {dict(zip(self.data['names'].values(), counts))}")
```
[the remark this answers is not in the document]
Comment (Kag-Bar) [reply ultralytics/models/yolo/classify/train.py:141]: fixed

File: ultralytics/utils/loss.py [hunk 5]
```diff
@@ -964,12 +964,196 @@ def calculate_keypoints_loss(
...
+
+        if cls_loss == "cb_focal" and class_counts is not None:
+            # Effective-number weighting: w_c = (1 - β) / (1 - β^n_c)
+            counts = torch.tensor(class_counts, dtype=torch.float32)
+            effective_num = 1.0 - self.cb_beta**counts
+            cb_weights = (1.0 - self.cb_beta) / effective_num
+            cb_weights = cb_weights / cb_weights.sum() * len(cb_weights)  # normalize so mean ≈ 1
+            if importance is not None:
+                cb_weights = cb_weights * importance  # combine with user importance
+            self._weight = cb_weights
```
[the remark this answers is not in the document]
Comment (Kag-Bar) [reply ultralytics/utils/loss.py:1047]: fixed

File: ultralytics/models/yolo/classify/train.py [hunk 6]
```diff
@@ -65,8 +65,80 @@ def __init__(self, cfg=DEFAULT_CFG, overrides: dict[str, Any] | None = None, _ca
...
-        """Set the YOLO model's class names from the loaded dataset."""
+        """Set the YOLO model's class names and classification loss config from the loaded dataset."""
         self.model.names = self.data["names"]
+        self.model.args = self.args  # attach hyperparameters so init_criterion() can read them
+
+        # Resolve class_weights (same logic as DetectionTrainer)
+        self._resolve_class_weights()
+
+        # Compute per-class sample counts from training set for class-balanced losses
+        self._compute_class_counts()
```
Comment (LeDuySon) [inline ultralytics/models/yolo/classify/train.py:76]: can we move tje resolve classs and compute class outside of this function because i think the role of them is different

**Review by LeDuySon — CHANGES_REQUESTED** [review]

File: ultralytics/nn/modules/head.py [hunk 7]
```diff
@@ -793,6 +794,7 @@ class Classify(nn.Module):
     """
 
     export = False  # export mode
+    _return_features = False  # set True for ArcFace (loss computes its own angular logits)
```
Comment (LeDuySon) [inline ultralytics/nn/modules/head.py:797]: How can we set this to True from config file, i dont see any part related to this

File: ultralytics/models/yolo/classify/train.py (hunk 6 above)
Comment (Kag-Bar) [reply ultralytics/models/yolo/classify/train.py:76]: How come the role is different?

File: ultralytics/nn/modules/head.py (hunk 7 above)
Comment (Kag-Bar) [reply ultralytics/nn/modules/head.py:797]: `_return_features `is a private internal flag (hence the _ prefix) — it is not user-facing and does not need to be set from config.
It is automatically managed by ClassificationModel.init_criterion() in nn/tasks.py (lines 719–738):
When the user sets cls_loss=arcface → init_criterion() sets head._return_features = True and passes head.linear.weight to the loss.
When cls_loss is anything else (or switches away from ArcFace) → it resets head._return_features = False.
So the only config knob the user touches is cls_loss. The flag is toggled as a side-effect of that choice, and unit tests 20–22 in tests/test_classification_loss.py explicitly verify this toggle behavior (including the ArcFace → CE switch and the no-args fallback).

File: ultralytics/models/yolo/classify/train.py (hunk 6 above)
Comment (Kag-Bar) [reply ultralytics/models/yolo/classify/train.py:76]: I dont want to compute it outside, i think it is a potential issue of values propegation, and is common to compute class balance in the module to extract the classes weights

Comment (LeDuySon) [reply ultralytics/models/yolo/classify/train.py:76]: i mean the function named is set_model_attributes and the two functions you add inside not related to set model attributes

Comment (Kag-Bar) [reply ultralytics/models/yolo/classify/train.py:76]: but in those functions we are setting `self.args.class_weights_resolved` and `self.args.class_counts = counts` which are part of the self model argumants (e.g. attributes), no? or i still dont get it

**Comment by LeDuySon** [comment]
Lgtm

File: ultralytics/models/yolo/classify/train.py (hunk 6 above)
Comment (LeDuySon) [reply ultralytics/models/yolo/classify/train.py:76]: Oh really, these are belong to model? I dont get the full context, you might right

**Review by LeDuySon — APPROVED** [review]
